File: src/backend/engine/setups.py
"""Rilevamento Setup A (pullback in trend) e Setup B (breakout da compressione).



Implementazione fedele a docs/STRATEGIA_SWING.md §4-§6. Output: dict con livelli

operativi (trigger, stop) da verificare manualmente su TradingView prima dell'ordine.

"""

import pandas as pd



from config import MARKET_PARAMS, MAX_STOP_ATR, RVOL_BREAKOUT

from engine.indicators import atr, bollinger_width, ema, rsi, rvol



SQUEEZE_LOOKBACK = 60

RANGE_BARS = 15  # ~3 settimane di borsa



# Fallback storici (market=None): stessi valori usati prima della Fase 5.

_DEFAULT_PARAMS = {

    "RANGE_BARS": RANGE_BARS,

    "SQUEEZE_LOOKBACK": SQUEEZE_LOOKBACK,

    "RSI_LONG_MIN": 40,

    "RSI_SHORT_MAX": 60,

}





def _market_params(market: str | None) -> dict:

    """Parametri per mercato da MARKET_PARAMS, vecchi default come fallback."""

    if market in MARKET_PARAMS:

        return {**_DEFAULT_PARAMS, **MARKET_PARAMS[market]}

    return _DEFAULT_PARAMS





def _round_px(x: float) -> float:

    return float(f"{x:.6g}")





def setup_a_metrics(df: pd.DataFrame, direction: str, market: str | None = None) -> dict | None:

    """Metriche e flag Setup A — usate da detect_setup_a e diagnostica."""

    if len(df) < 220:

        return None

    p = _market_params(market)

    close, volume = df["close"], df["volume"]

    e20, e50, e200 = ema(close, 20), ema(close, 50), ema(close, 200)

    a = float(atr(df).iloc[-1])

    r = float(rsi(close).iloc[-1])

    last = float(close.iloc[-1])

    vol5 = float(volume.rolling(5).mean().iloc[-1])

    vol20 = float(volume.rolling(20).mean().iloc[-1])

    vol_declining = vol5 < vol20



    if direction == "long":

        aligned = (

            e20.iloc[-1] > e50.iloc[-1] > e200.iloc[-1] and e50.iloc[-1] > e50.iloc[-6]

        )

        in_zone = (e50.iloc[-1] - 0.5 * a) <= last <= (e20.iloc[-1] + 0.25 * a)

        momentum_ok = r > p["RSI_LONG_MIN"]

        swing = float(df["low"].rolling(10).min().iloc[-1])

        stop = swing - 0.5 * a

        trigger = float(df["high"].iloc[-2:].max())

    else:

        aligned = (

            e20.iloc[-1] < e50.iloc[-1] < e200.iloc[-1] and e50.iloc[-1] < e50.iloc[-6]

        )

        in_zone = (e20.iloc[-1] - 0.25 * a) <= last <= (e50.iloc[-1] + 0.5 * a)

        momentum_ok = r < p["RSI_SHORT_MAX"]

        swing = float(df["high"].rolling(10).max().iloc[-1])

        stop = swing + 0.5 * a

        trigger = float(df["low"].iloc[-2:].min())



    stop_dist = abs(trigger - stop)

    stop_geometry_ok = stop_dist <= MAX_STOP_ATR * a



    return {

        "aligned": aligned,

        "in_zone": in_zone,

        "momentum_ok": momentum_ok,

        "vol_declining": vol_declining,

        "stop_geometry_ok": stop_geometry_ok,

        "rsi": r,

        "atr": a,

        "trigger": trigger,

        "stop": stop,

        "stop_dist": stop_dist,

        "vol5": vol5,

        "vol20": vol20,

    }





def setup_b_metrics(df: pd.DataFrame, direction: str, market: str | None = None) -> dict | None:

    """Metriche e flag Setup B — usate da detect_setup_b e diagnostica."""

    if len(df) < 220:

        return None

    p = _market_params(market)

    squeeze_lookback = p["SQUEEZE_LOOKBACK"]

    range_bars = p["RANGE_BARS"]

    close = df["close"]

    e200 = float(ema(close, 200).iloc[-1])

    last = float(close.iloc[-1])

    a = float(atr(df).iloc[-1])



    bbw = bollinger_width(close)

    bbw_last = float(bbw.iloc[-1])

    bbw_thresh = float(bbw.iloc[-squeeze_lookback:].quantile(0.10))

    squeeze = bbw_last <= bbw_thresh



    # Il range di compressione esclude la barra corrente: includerla rendeva
    # impossibile close > rng_high (close <= high), quindi breakout_triggered
    # era sempre False (codice morto). Il breakout è la rottura, da parte della
    # barra corrente, del range delle RANGE_BARS barre precedenti.
    rng_high = float(df["high"].iloc[-range_bars - 1:-1].max())

    rng_low = float(df["low"].iloc[-range_bars - 1:-1].min())



    if direction == "long":

        context_ok = last >= e200

        trigger = rng_high

        stop = max(trigger - a, rng_low)

    else:

        context_ok = last <= e200

        trigger = rng_low

        stop = min(trigger + a, rng_high)



    stop_dist = abs(trigger - stop)

    stop_geometry_ok = stop_dist <= MAX_STOP_ATR * a



    rv_series = rvol(df["volume"])

    rv = float(rv_series.iloc[-1]) if pd.notna(rv_series.iloc[-1]) else 0.0

    triggered = (

        direction == "long" and last > trigger and rv >= RVOL_BREAKOUT

    ) or (

        direction == "short" and last < trigger and rv >= RVOL_BREAKOUT

    )



    return {

        "squeeze": squeeze,

        "context_ok": context_ok,

        "stop_geometry_ok": stop_geometry_ok,

        "breakout_triggered": triggered,

        "bbw_last": bbw_last,

        "bbw_thresh": bbw_thresh,

        "atr": a,

        "trigger": trigger,

        "stop": stop,

        "stop_dist": stop_dist,

        "rvol": rv,

        "e200": e200,

        "last": last,

    }





def detect_setup_a(df: pd.DataFrame, direction: str, market: str | None = None) -> dict | None:

    """Pullback verso EMA20-EMA50 in trend allineato, volume in calo, RSI intatto."""

    m = setup_a_metrics(df, direction, market)

    if m is None:

        return None

    if not (m["aligned"] and m["in_zone"] and m["momentum_ok"] and m["vol_declining"]):

        return None

    if not m["stop_geometry_ok"]:

        return None



    return {

        "setup": "A",

        "direction": direction,

        "entry_trigger": _round_px(m["trigger"]),

        "stop": _round_px(m["stop"]),

        "atr": _round_px(m["atr"]),

        "rsi": round(m["rsi"], 1),

        "note": "Pullback in trend: conferma il trigger sulla candela 4H con volume (TradingView)",

    }





def detect_setup_b(df: pd.DataFrame, direction: str, market: str | None = None) -> dict | None:

    """Compressione di volatilità (squeeze BB / range) + livello di rottura con RVOL>=2."""

    m = setup_b_metrics(df, direction, market)

    if m is None:

        return None

    if not m["squeeze"]:

        return None

    if not m["context_ok"]:

        return None

    if not m["stop_geometry_ok"]:

        return None



    return {

        "setup": "B",

        "direction": direction,

        "entry_trigger": _round_px(m["trigger"]),

        "stop": _round_px(m["stop"]),

        "atr": _round_px(m["atr"]),

        "status_hint": "triggered" if m["breakout_triggered"] else "watch",

        "note": f"Breakout da squeeze: serve chiusura daily oltre il livello con RVOL>={RVOL_BREAKOUT}",

    }





def trigger_status_4h(df_4h: pd.DataFrame, direction: str, trigger: float) -> str:

    """'triggered' | 'near' | 'watch' valutato sulle candele 4H (con conferma volume)."""

    if df_4h.empty or len(df_4h) < 25:

        return "watch"

    last_close = float(df_4h["close"].iloc[-1])

    vol_ok = df_4h["volume"].iloc[-1] > df_4h["volume"].rolling(20).mean().iloc[-1]

    if direction == "long":

        if last_close > trigger and vol_ok:

            return "triggered"

        if last_close >= trigger * 0.99:

            return "near"

    else:

        if last_close < trigger and vol_ok:

            return "triggered"

        if last_close <= trigger * 1.01:

            return "near"

    return "watch"

File: vision-mobile/src/engine/setups.ts
/** Rilevamento Setup A (pullback in trend) e Setup B (breakout da compressione).
 *
 * Implementazione fedele a docs/STRATEGIA_SWING.md §4-§6.
 */
import { MARKET_PARAMS, MAX_STOP_ATR, RVOL_BREAKOUT, type MarketParams } from "../config";
import { atr, bollingerWidth, ema, rsi, rvol } from "./indicators";
import type { OHLCVBar } from "./types";

export const SQUEEZE_LOOKBACK = 60;
export const RANGE_BARS = 15;

// Fallback storici (market assente): stessi valori usati prima della Fase 5.
const DEFAULT_PARAMS: MarketParams = {
  RANGE_BARS,
  SQUEEZE_LOOKBACK,
  RSI_LONG_MIN: 40,
  RSI_SHORT_MAX: 60,
};

export function marketParams(market?: "crypto" | "stocks" | null): MarketParams {
  if (market && MARKET_PARAMS[market]) {
    return { ...DEFAULT_PARAMS, ...MARKET_PARAMS[market] };
  }
  return DEFAULT_PARAMS;
}

function roundPx(x: number): number {
  return parseFloat(x.toPrecision(6));
}

function rollingMean(arr: number[], window: number): number[] {
  const out = new Array<number>(arr.length).fill(NaN);
  for (let i = window - 1; i < arr.length; i++) {
    const slice = arr.slice(i - window + 1, i + 1);
    out[i] = slice.reduce((a, b) => a + b, 0) / window;
  }
  return out;
}

function rollingMin(arr: number[], window: number): number[] {
  const out = new Array<number>(arr.length).fill(NaN);
  for (let i = window - 1; i < arr.length; i++) {
    out[i] = Math.min(...arr.slice(i - window + 1, i + 1));
  }
  return out;
}

function rollingMax(arr: number[], window: number): number[] {
  const out = new Array<number>(arr.length).fill(NaN);
  for (let i = window - 1; i < arr.length; i++) {
    out[i] = Math.max(...arr.slice(i - window + 1, i + 1));
  }
  return out;
}

function quantile(arr: number[], q: number): number {
  const sorted = [...arr].filter((v) => Number.isFinite(v)).sort((a, b) => a - b);
  if (!sorted.length) return NaN;
  const pos = (sorted.length - 1) * q;
  const lo = Math.floor(pos);
  const hi = Math.ceil(pos);
  if (lo === hi) return sorted[lo];
  return sorted[lo] + (sorted[hi] - sorted[lo]) * (pos - lo);
}

function ohlcvCols(bars: OHLCVBar[]) {
  return {
    open: bars.map((b) => b.open),
    high: bars.map((b) => b.high),
    low: bars.map((b) => b.low),
    close: bars.map((b) => b.close),
    volume: bars.map((b) => b.volume),
  };
}

export interface SetupAMetrics {
  aligned: boolean;
  in_zone: boolean;
  momentum_ok: boolean;
  vol_declining: boolean;
  stop_geometry_ok: boolean;
  rsi: number;
  atr: number;
  trigger: number;
  stop: number;
  stop_dist: number;
  vol5: number;
  vol20: number;
}

export function setupAMetrics(
  bars: OHLCVBar[],
  direction: string,
  market?: "crypto" | "stocks" | null
): SetupAMetrics | null {
  if (bars.length < 220) return null;
  const p = marketParams(market);

  const { high, low, close, volume } = ohlcvCols(bars);
  const e20 = ema(close, 20);
  const e50 = ema(close, 50);
  const e200 = ema(close, 200);
  const atrSeries = atr(high, low, close);
  const rsiSeries = rsi(close);
  const a = atrSeries[atrSeries.length - 1];
  const r = rsiSeries[rsiSeries.length - 1];
  const last = close[close.length - 1];
  const vol5Series = rollingMean(volume, 5);
  const vol20Series = rollingMean(volume, 20);
  const vol5 = vol5Series[vol5Series.length - 1];
  const vol20 = vol20Series[vol20Series.length - 1];
  const volDeclining = vol5 < vol20;

  let aligned: boolean;
  let inZone: boolean;
  let momentumOk: boolean;
  let swing: number;
  let stop: number;
  let trigger: number;

  if (direction === "long") {
    aligned =
      e20[e20.length - 1] > e50[e50.length - 1] &&
      e50[e50.length - 1] > e200[e200.length - 1] &&
      e50[e50.length - 1] > e50[e50.length - 6];
    inZone = e50[e50.length - 1] - 0.5 * a <= last && last <= e20[e20.length - 1] + 0.25 * a;
    momentumOk = r > p.RSI_LONG_MIN;
    const swingSeries = rollingMin(low, 10);
    swing = swingSeries[swingSeries.length - 1];
    stop = swing - 0.5 * a;
    trigger = Math.max(high[high.length - 2], high[high.length - 1]);
  } else {
    aligned =
      e20[e20.length - 1] < e50[e50.length - 1] &&
      e50[e50.length - 1] < e200[e200.length - 1] &&
      e50[e50.length - 1] < e50[e50.length - 6];
    inZone = e20[e20.length - 1] - 0.25 * a <= last && last <= e50[e50.length - 1] + 0.5 * a;
    momentumOk = r < p.RSI_SHORT_MAX;
    const swingSeries = rollingMax(high, 10);
    swing = swingSeries[swingSeries.length - 1];
    stop = swing + 0.5 * a;
    trigger = Math.min(low[low.length - 2], low[low.length - 1]);
  }

  const stopDist = Math.abs(trigger - stop);
  const stopGeometryOk = stopDist <= MAX_STOP_ATR * a;

  return {
    aligned,
    in_zone: inZone,
    momentum_ok: momentumOk,
    vol_declining: volDeclining,
    stop_geometry_ok: stopGeometryOk,
    rsi: r,
    atr: a,
    trigger,
    stop,
    stop_dist: stopDist,
    vol5,
    vol20,
  };
}

export interface SetupBMetrics {
  squeeze: boolean;
  context_ok: boolean;
  stop_geometry_ok: boolean;
  breakout_triggered: boolean;
  bbw_last: number;
  bbw_thresh: number;
  atr: number;
  trigger: number;
  stop: number;
  stop_dist: number;
  rvol: number;
  e200: number;
  last: number;
}

export function setupBMetrics(
  bars: OHLCVBar[],
  direction: string,
  market?: "crypto" | "stocks" | null
): SetupBMetrics | null {
  if (bars.length < 220) return null;
  const p = marketParams(market);

  const { high, low, close, volume } = ohlcvCols(bars);
  const e200Series = ema(close, 200);
  const e200 = e200Series[e200Series.length - 1];
  const last = close[close.length - 1];
  const atrSeries = atr(high, low, close);
  const a = atrSeries[atrSeries.length - 1];

  const bbw = bollingerWidth(close);
  const bbwLast = bbw[bbw.length - 1];
  // Quantile sulle barre PRECEDENTI la corrente — speculare a setups.py.
  const bbwSlice = bbw.slice(-p.SQUEEZE_LOOKBACK - 1, -1);
  const bbwThresh = quantile(bbwSlice, 0.1);
  const squeeze = bbwLast <= bbwThresh;

  // Il range di compressione esclude la barra corrente: includerla rendeva
  // impossibile close > rangeHigh (close <= high), quindi breakout_triggered
  // era sempre false (codice morto). Speculare al fix nel backend Python.
  const rangeHigh = Math.max(...high.slice(-p.RANGE_BARS - 1, -1));
  const rangeLow = Math.min(...low.slice(-p.RANGE_BARS - 1, -1));

  let contextOk: boolean;
  let trigger: number;
  let stop: number;

  if (direction === "long") {
    contextOk = last >= e200;
    trigger = rangeHigh;
    stop = Math.max(trigger - a, rangeLow);
  } else {
    contextOk = last <= e200;
    trigger = rangeLow;
    stop = Math.min(trigger + a, rangeHigh);
  }

  const stopDist = Math.abs(trigger - stop);
  const stopGeometryOk = stopDist <= MAX_STOP_ATR * a;

  const rvSeries = rvol(volume);
  const rvVal = rvSeries[rvSeries.length - 1];
  const rv = rvVal != null && !Number.isNaN(rvVal) ? rvVal : 0;
  const triggered =
    (direction === "long" && last > trigger && rv >= RVOL_BREAKOUT) ||
    (direction === "short" && last < trigger && rv >= RVOL_BREAKOUT);

  return {
    squeeze,
    context_ok: contextOk,
    stop_geometry_ok: stopGeometryOk,
    breakout_triggered: triggered,
    bbw_last: bbwLast,
    bbw_thresh: bbwThresh,
    atr: a,
    trigger,
    stop,
    stop_dist: stopDist,
    rvol: rv,
    e200,
    last,
  };
}

export interface SetupSignal {
  setup: "A" | "B";
  direction: string;
  entry_trigger: number;
  stop: number;
  atr: number;
  rsi?: number;
  status_hint?: "triggered" | "watch";
  note: string;
}

export function detectSetupA(
  bars: OHLCVBar[],
  direction: string,
  market?: "crypto" | "stocks" | null
): SetupSignal | null {
  const m = setupAMetrics(bars, direction, market);
  if (m == null) return null;
  if (!(m.aligned && m.in_zone && m.vol_declining)) return null;
  if (!m.stop_geometry_ok) return null;

  return {
    setup: "A",
    direction,
    entry_trigger: roundPx(m.trigger),
    stop: roundPx(m.stop),
    atr: roundPx(m.atr),
    rsi: Math.round(m.rsi * 10) / 10,
    note: "Pullback in trend: verifica chiusura oltre il livello di rottura su 4H con volume (TradingView)",
  };
}

export function detectSetupB(
  bars: OHLCVBar[],
  direction: string,
  market?: "crypto" | "stocks" | null
): SetupSignal | null {
  const m = setupBMetrics(bars, direction, market);
  if (m == null) return null;
  if (!m.squeeze) return null;
  if (!m.context_ok) return null;
  if (!m.stop_geometry_ok) return null;

  return {
    setup: "B",
    direction,
    entry_trigger: roundPx(m.trigger),
    stop: roundPx(m.stop),
    atr: roundPx(m.atr),
    status_hint: m.breakout_triggered ? "triggered" : "watch",
    note: `Breakout da squeeze: serve chiusura daily oltre il livello con RVOL>=${RVOL_BREAKOUT}`,
  };
}

export function triggerStatus4h(
  bars4h: OHLCVBar[],
  direction: string,
  trigger: number
): "triggered" | "near" | "watch" {
  if (!bars4h.length || bars4h.length < 25) return "watch";

  const lastClose = bars4h[bars4h.length - 1].close;
  const vol = bars4h.map((b) => b.volume);
  // Media delle 20 barre precedenti (equivalente a shift(1) in pandas).
  const volMean = rollingMean(vol, 20);
  const volOk = vol[vol.length - 1] > volMean[volMean.length - 2];

  if (direction === "long") {
    if (lastClose > trigger && volOk) return "triggered";
    if (lastClose >= trigger * 0.99) return "near";
  } else {
    if (lastClose < trigger && volOk) return "triggered";
    if (lastClose <= trigger * 1.01) return "near";
  }
  return "watch";
}
